cmd/backup.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/spf13/cobra"
+	"github.com/tigrisdata/tigris-cli/client"
+	"github.com/tigrisdata/tigris-cli/util"
+	"github.com/tigrisdata/tigris-client-go/driver"
+)
+
+func listDatabases(ctx context.Context, dbFilter []string) ([]string, error) {
+	databases := []string{}
+
+	resp, err := client.Get().ListDatabases(ctx)
+	if err != nil {
+		return databases, util.Error(err, "list databases")
+	}
+
+	for _, v := range resp {
+		if len(dbFilter) == 0 || util.Contains(dbFilter, v) {
+			databases = append(databases, v)
+		}
+	}
+
+	return databases, nil
+}
+
+func listCollections(ctx context.Context, dbName string, collectionFilter []string) ([]string, error) {
+	collections := []string{}
+
+	resp, err := client.Get().UseDatabase(dbName).ListCollections(ctx)
+	if err != nil {
+		return collections, util.Error(err, "list collections")
+	}
+
+	for _, v := range resp {
+		if len(collectionFilter) == 0 || util.Contains(collectionFilter, v) {
+			collections = append(collections, v)
+		}
+	}
+
+	return collections, nil
+}
+
+var backupCmd = &cobra.Command{
+	Use:   "backup [filters]",
+	Short: "Dumps documents and schemas to JSON files",
+	Long: `Dumps documents and schemas to JSON files
+	in the current working directory.
+	
+	If a database filter is provided it only dumps the schemas of the databases specified.
+	Likewise, collection filters will limit the output to matching collection names.`,
+	Args: cobra.MinimumNArgs(0),
+	Run: func(cmd *cobra.Command, args []string) {
+		withLogin(cmd.Context(), func(ctx context.Context) error {
+			dbFilter := []string{}
+			collectionFilter := []string{}
+
+			ctx, cancel := context.WithTimeout(ctx, 1*time.Hour)
+			defer cancel()
+			os.Setenv("TIGRIS_TIMEOUT", "1h")
```

Comment:
this is not needed so as you create new context with 1hour timeout above

Reply:
Yeah I didn't intend to leave this in but context wasn't working. Unless I set this in the invoking shell it still isn't working right.

Reply:
```
❯ export TIGRIS_TIMEOUT=1h

❯ /Users/rbarabas/src/int/tigris-cli/tigris backup
 [.] ./tigris_netlify_starter.schema
 [*] ./tigris_netlify_starter.todoItems.json
 [.] ./ecommerce.schema
 [*] ./ecommerce.users.json
 [.] ./meanStackExample.schema
 [*] ./meanStackExample.employees.json
 [.] ./todoStarterApp.schema
 [*] ./todoStarterApp.todoItems.json
 [.] ./meteorites.schema
 [*] ./meteorites.landings.json
 [.] ./shop.schema
 [*] ./shop.products.json
 [*] ./shop.users.json
 [*] ./shop.orders.json
 [.] ./ecommerce_db.schema
 [*] ./ecommerce_db.ecommerce_collection.json
 [.] ./ycsb_tigris.schema
 [*] ./ycsb_tigris.user_tables.json
 [.] ./hello_tigris.schema
 [*] ./hello_tigris.user.json
 [.] ./tigris_vercel_starter.schema
 [*] ./tigris_vercel_starter.todoItems.json
 [.] ./catalog.schema
 [*] ./catalog.products.json
 [.] ./newchic.schema
 [*] ./newchic.bags.json
 [*] ./newchic.beauty.json
 [*] ./newchic.jewelry.json
 [*] ./newchic.kids.json
 [*] ./newchic.men.json
 [*] ./newchic.shoes.json
 [*] ./newchic.women.json
 [*] ./newchic.accessories.json
 [.] ./db2.schema
 [.] ./testdb.schema
 [*] ./testdb.users.json
 [.] ./home_depot.schema
 [.] ./tigris_starter_ts.schema
 [*] ./tigris_starter_ts.orders.json
 [*] ./tigris_starter_ts.products.json
 [*] ./tigris_starter_ts.social_messages.json
 [*] ./tigris_starter_ts.user_events.json
 [*] ./tigris_starter_ts.users.json
 [.] ./tigris_starter_java.schema
 [*] ./tigris_starter_java.orders.json
 [*] ./tigris_starter_java.product_collection.json
 [*] ./tigris_starter_java.users.json
 [.] ./auth0.schema
 [*] ./auth0.users.json
 [.] ./db1.schema
 [*] ./db1.todoList.json
 [*] ./db1.u12.json
 [*] ./db1.t1.json
 [*] ./db1.test_nested_objects_arrays.json
 [*] ./db1.u14.json
 [*] ./db1.u2.json
 [*] ./db1.u16.json
 [*] ./db1.coll1.json
 [*] ./db1.users1.json
 [*] ./db1.u15.json
 [*] ./db1.col16.json
 [*] ./db1.coll111.json
 [.] ./jmj_test.schema
 [*] ./jmj_test.foo.json
 ```

Reply:
I removed the context and left the `env` based solution. It seems to be working with that:

```
❯ /Users/rbarabas/src/int/tigris-cli/tigris backup
 [o] using timeout 1h
 [.] ./meanStackExample.schema
 [*] ./meanStackExample.employees.json
 [.] ./db2.schema
 [.] ./tigris_starter_ts.schema
 [*] ./tigris_starter_ts.products.json
 [*] ./tigris_starter_ts.social_messages.json
 [*] ./tigris_starter_ts.user_events.json
 [*] ./tigris_starter_ts.users.json
 [*] ./tigris_starter_ts.orders.json
 [.] ./newchic.schema
 [*] ./newchic.kids.json
 [*] ./newchic.men.json
 [*] ./newchic.shoes.json
 [*] ./newchic.women.json
 [*] ./newchic.accessories.json
 [*] ./newchic.bags.json
 [*] ./newchic.beauty.json
 [*] ./newchic.jewelry.json
 [.] ./auth0.schema
 [*] ./auth0.users.json
 [.] ./hello_tigris.schema
 [*] ./hello_tigris.user.json
 [.] ./tigris_vercel_starter.schema
 [*] ./tigris_vercel_starter.todoItems.json
 [.] ./ecommerce_db.schema
 [*] ./ecommerce_db.ecommerce_collection.json
 [.] ./tigris_netlify_starter.schema
 [*] ./tigris_netlify_starter.todoItems.json
 [.] ./tigris_starter_java.schema
 [*] ./tigris_starter_java.product_collection.json
 [*] ./tigris_starter_java.users.json
 [*] ./tigris_starter_java.orders.json
 [.] ./db1.schema
 [*] ./db1.col16.json
 [*] ./db1.coll1.json
 [*] ./db1.u15.json
 [*] ./db1.users1.json
 [*] ./db1.u16.json
 [*] ./db1.u2.json
 [*] ./db1.coll111.json
 [*] ./db1.t1.json
 [*] ./db1.todoList.json
 [*] ./db1.test_nested_objects_arrays.json
 [*] ./db1.u14.json
 [*] ./db1.u12.json
 [.] ./testdb.schema
 [*] ./testdb.users.json
 [.] ./todoStarterApp.schema
 [*] ./todoStarterApp.todoItems.json
 [.] ./jmj_test.schema
 [*] ./jmj_test.foo.json
 [.] ./ecommerce.schema
 [*] ./ecommerce.users.json
 [.] ./shop.schema
 [*] ./shop.orders.json
 [*] ./shop.products.json
 [*] ./shop.users.json
 [.] ./catalog.schema
 [*] ./catalog.products.json
 [.] ./ycsb_tigris.schema
 [*] ./ycsb_tigris.user_tables.json
 [.] ./home_depot.schema
 [.] ./meteorites.schema
 [*] ./meteorites.landings.json
 ```

Reply:
It's better to make context work.
I think I know what is the problem:
```
ctx, cancel := context.WithTimeout(ctx, 1*time.Hour)
```
should be:
```
ctx, cancel := context.WithTimeout(context.Background(), 1*time.Hour)
```

Reply:
So loose the `ctx` passed by cobra and just initialize clean context? 

I presume this also means we would want to "parse" the timeout specification.

Reply:
Right, loose cobra's ctx.
Just add a constant in the code for backup timeout for now.
Later we can introduce config variable.

Reply:
OK!

Reply:
It worked, no context timeout. We can close this. I kept the configurability just limited the unit to seconds for now.